sending sleep command also when the error is unrecoverable

onicolaos opened the pull request:
The chip will be occasionally stuck in a state where it will not respond. Sending a sleep command when an error seems to resolve the issue. An unrecoverable error before before retries have being exhausted seems to not have being covered with a sleep command previously

onicolaos committed:
add flag for sending sleep command for recovery also when the error is unrecoverable

The chip will be occasionally stuck in a state where it will not respond. Sending a sleep command when an error seems to resolve the issue. An unrecoverable error before before retries have being exhausted seems to not have being covered with a sleep command previously

onicolaos committed:
idling without a successful response

"idle" command should be sent if returning a successful response or if not sending a sleep command

codearranger commented on src/ecc.rs:
```diff
@@ -274,24 +276,34 @@ impl Ecc {
         }
+
+        if should_sleep {
+        self.transport.send_sleep(); // Send the sleep command
```
You might want to set should_sleep = false in here somewhere.

onicolaos committed:
reset the flag

Reset the flag after sending the sleep command

onicolaos committed:
revert should_sleep and add error mitigation as last resort

mawdegroot commented on src/ecc.rs:
```diff
@@ -283,15 +283,36 @@ impl Ecc {
+                }
             }
         }
+        self.error_mitigation();
```
With this addition the sleep at line 278 can also be removed since that will fall through to here.

onicolaos committed:
make use of thread and pass wake_delay as an argument

onicolaos committed:
remove unnecessary sleep in mid loop

onicolaos committed:
fix off by one and dump the mitigation

fix the off by ones to make the retry/break logic sane again
dump the mitigation and replace it by normal sleeps (which didn't fire before so they probably just work)
only idle on success

onicolaos committed:
wait longer on last resort

do a last resort longer wait on the last retry in case the ecc is seriously slow to respond

onicolaos committed:
small fix and cleanup
